bentoml/_internal/models/model.py

```diff
@@ -0,0 +1,257 @@
+        if self._info is not None:
+            return self._info
+
+        with self._fs.open(MODEL_YAML_FILENAME, "r") as model_yaml:
+            self._info = ModelInfo.from_yaml_file(model_yaml)
+            return self._info
+
+    @property
+    def creation_time(self) -> datetime:
+        return self.info["creation_time"]
+
+    def export(self, path: str):
+        out_fs = fs.open_fs(path, create=True, writeable=True)
+        fs.mirror.mirror(self._fs, out_fs, copy_if_newer=False)
+
+    def push(self):
+        pass
+
+    def load_runner(self) -> "Runner":
+        raise NotImplementedError
+
+    def validate(self):
+        return self._fs.isfile(MODEL_YAML_FILENAME)
+
+    def __str__(self):
+        return f'Bento(tag="{self.tag}", path="{self.path}")'
+
+
+class SysPathModel(Model):
+    @staticmethod
+    def from_Model(model: Model) -> "SysPathModel":
+        try:
+            model._fs.getsyspath("/")
+        except fs.errors.NoSysPath:
+            raise TypeError(f"{model} is not in the OS filesystem")
+        model.__class__ = SysPathModel
+        return model  # type: ignore
+
+    @property
+    def path(self) -> str:
+        return self.path_of("/")
+
+    def path_of(self, item: str) -> str:
+        return self._fs.getsyspath(item)
+
+    def save(
+        self, model_store: "ModelStore" = Provide[BentoMLContainer.model_store]
+    ) -> "SysPathModel":
+        if self._info is not None:
+            with self._fs.open(MODEL_YAML_FILENAME, "w") as model_yaml:
+                self._info.dump(model_yaml)
+
+        if not self.validate():
+            logger.warning(f"Failed to create Model for {self.tag}, not saving.")
+            raise BentoMLException("Failed to save Model because it was invalid")
+
+        with model_store.register(self.tag) as model_path:
+            os.rmdir(model_path)
+            shutil.move(self._fs.getsyspath("/"), model_path)
+            self._fs.close()
+            self._fs = fs.open_fs(model_path)
+
+        return self
+
+
+class ModelStore(Store[SysPathModel]):
+    def __init__(self, base_path: t.Union[PathType, FS]):
+        super().__init__(base_path, SysPathModel)
+
+
+if t.TYPE_CHECKING:
+    ModelInfoBase = UserDict[str, t.Any]
+else:
+    ModelInfoBase = UserDict
+
+
+@attr.define(repr=False)
+class ModelInfo(ModelInfoBase):
+    tag: Tag
+    module: str
+    labels: t.Dict[str, t.Any]
+    options: t.Dict[str, t.Any]
+    metadata: t.Dict[str, t.Any]
+    context: t.Dict[str, t.Any]
+    bentoml_version: str = BENTOML_VERSION
+    api_version: str = "v1"
+    creation_time: datetime = attr.field(factory=lambda: datetime.now(timezone.utc))
+
+    def __attrs_post_init__(self):
+        self.validate()
+
+    def dump(self, stream: t.IO[t.Any]):
+        return yaml.dump(self, stream, sort_keys=False)
+
+    @classmethod
+    def from_yaml_file(cls, stream: t.IO[t.Any]):
+        try:
+            yaml_content = yaml.safe_load(stream)
```

Comment:
Note: `safe_load` isn't able to load a decent amount of yaml objects that we're dumping, but I'm not what the security model should be here. We might want to support custom save/load routines for metadata or just require that all the relevant data is `str`.